Differential Team into master

rlee287 opened the pull request.

pluto-de committed:
Update queries.js

sspathak committed:
fixed errors due to incorrect quotes

sspathak committed:
fixed scenarioExists and scenarioPageExists to use promises

sspathak committed:
removed temporary test code

rlee287 committed:
Add body text to getPage queries and default status

rlee287 committed:
Add back .env file example comment to queries.js

rlee287 committed:
Manually copy queries.js changeset from 8d1fbde

rlee287 committed:
Adjust getReflectPage functions based on frontend-integration stuffs

rlee287 committed:
Add missing commas to map object response builder

rlee287 committed:
queries.js query typofix argh

rlee287 committed:
Colon period typo fix arghghghghgh

rlee287 committed:
Add a getConclusionPage function

rlee287 committed:
For getBlahReflectPage, callback on empty object if SQL returns nothing

rlee287 committed:
results.rows.length instead of results.length

(cherry picked from commit 82513b5bcc211bb5356dd15fa3a4f78342b41041)

rlee287 committed:
Add function to set scenario status field

Assuming that server backend does all the value checks first

rlee287 committed:
Ensure that promptNum is passed in when adding in reflection responses

rlee287 committed:
Give all reflect responses a promptNum arg (and not just init)

kungoyal committed:
Updated env config

kungoyal committed:
Fixed ON CONFLICT error

(cherry picked from commit f6fb9452fe2e6264dc516e83569bfe79e660bebf)

sspathak committed:
added function skeleton for version control (csv)

rlee287 committed:
Reflection response return should filter by submissions.scenario_id

(cherry picked from commit 0927d42cf93a1bc655a2b56ff0c00248c2710a63)

MarcotteCaleb committed:
Implemented createScenario

rlee287 committed:
Callback empty string for addReflectionResponse upon nonexistent ID

(cherry picked from commit ac3919d2806de5d0bdc0a56403bbe427195d3845)

rlee287 committed:
Merge branch 'master' into differential_team

rlee287 commented:
This can be merged either before or after the hotfix PR #208.

rasdy commented on goon-universe/queries.js:
```diff
@@ -70,7 +76,17 @@ function getTaskPage(scenarioID, callback){
+    })
+}
+
+function getConclusionPage(scenarioID, callback){
```
Calling `GET /scenarios/conclusion` with `scenarioID: 1` returns a 404. Is the example data set up correctly to test this endpoint?

rlee287 replied:
Probably; I can double-check this once I update the example data

rlee287 replied:
Test data is not fully updated but this particular endpoint should work correctly now. Once you test this again we can resolve this conversation.

rlee287 replied:
@rasdy can you reverify when you have a chance?

rasdy replied:
Works fine. Ping me on slack next time I'm slow to respond on github

rasdy commented on goon-universe/queries.js:
```diff
@@ -241,35 +257,78 @@ function addFinalReflectResponse(studentID, input, promptNum, scenarioID, timest
         .then((succeeded) => callback(succeeded ? "Success!" : ""));
 }
 
+
+// scenarioExists(scenarioID)
+// .then(function(result){
+//     console.log(result);
+// })
+// .catch(function(err){
+//     console.log(err);
+// });
```
Are the commented-out functions like this left in on purpose?

rlee287 replied:
This comment explains how to use the scenarioExists function that is defined below.

rasdy commented on goon-universe/queries.js:
```diff
@@ -614,45 +689,25 @@ function getFinalReflectPage(scenarioID, callback){
-    pool.query(thisQuery, [scenarioID], (error,results) => {
-        if (error) {
+function getInitActionPageQuestionsAndChoices(scenarioID,questionID,callback){
+    let thisQuery='select question.question, question.id, mcq_option.option from pages, mcq, mcq_option, question where pages.order ='+ INIT_ACTION +'and scenario_id = $1 and mcq.page_id = question.mcq_id and pages.id=mcq.page_id and mcq_option.question_id = $2 and mcq_option.question_id = question.id'
+    pool.query(thisQuery, [scenarioID, questionID], (error,results) => {
+        if(error){
             throw error
         }
         callback(results.rows)
-    })  
+    })
 }
```
Why is `getInitActionPageQuestionsAndChoices()` taking in a questionID parameter, but also returing question IDs?

rlee287 replied:
Fixed in 05936594.

rasdy reviewed (changes requested):
Had a couple questions and brought up an issue. Someone from track 4 should review this as well, since I wouldn't know what to look for regarding the editor-related functions.

rlee287 committed:
Remove returning question ID when this is an input parameter

rlee287 committed:
Remove content field from conversation_task and mcq

Body text makes this redundant

rlee287 committed:
Add more checks to database setup

VARCHAR can be non-null empty string, so forbid NULL in those cases

CHECK("blah" <> '') means that field "blah" cannot be the empty string,
which is what would usually be intended for "required" string fields

rlee287 committed:
More database-level CHECK constraints

rlee287 committed:
First pass at updating sample data for testing purposes

kungoyal committed:
Merge

kungoyal committed:
Fixed some function signatures

kungoyal committed:
Fixed some function signatures

(cherry picked from commit a33843165747d54fa669ad6bb922b968c0954a40)

sspathak committed:
fixed editor functions' syntax

sspathak committed:
converted some editor functions to use async/await

sspathak committed:
added helper function for adding scenario pages, modified funcitons to use helper

rlee287 committed:
Add conversation count limit to conversation_task table

sspathak committed:
added a missing '}'

rlee287 committed:
Move conversation_limit to the correct table

rlee287 committed:
Update addConvTaskPage with conversation limit parameter

rlee287 committed:
Warning comment stating that addIntroPage needs fixing

rlee287 committed:
Add conv_limit to example data

The whole statement needs more fixing anyway

rlee287 commented:
Pinging @aj8uppal for code review later (not yet as we are still fixing some things)

rlee287 committed:
Fix conversation_limit positive value check

rlee287 committed:
Rename page_num to page_id and add REFERENCES constraint

rlee287 committed:
Return stakeholder designation as well

sspathak committed:
converted more editor functions to async/await

sspathak committed:
added MCQ functions for API

sspathak committed:
modified editor functions

aj8uppal reviewed (dismissed):
Looks good to me.

rasdy commented on goon-universe/app.js:
```diff
@@ -225,7 +225,7 @@ router.route('/scenarios/initialAction')
         }
         else{
-        db.getInitActions(scenarioID, function(result){
+        db.getInitActionPageQuestionsAndChoices(scenarioID, function(result){
```
Querying `GET /scenarios/initialAction` with `scenarioID: 1, studentID: 1` results in an unhandled exception being thrown.

kungoyal replied:
DT's implementation of the getInitActionPageQuestionsAndChoices takes questionID as an argument along with scenarioID. I believe this is the issue. Also this function does not require studentID. @rasdy @rlee287

rlee287 replied:
Function now does callback on empty object (if question and options do not exist) or on JSON object with `question` and `mcq_choices` fields. I will ping both of you on Slack once I finish these changes and push them.

rlee287 replied:
This should be fixed now.

rasdy commented on goon-universe/app.js:
```diff
@@ -252,7 +252,7 @@ router.route('/scenarios/initialAction')
         }
         else{
-        db.addInitActionChoice(studentID, scenarioID, data, function(result){
+        db.addInitActionResponse(studentID, scenarioID, data, function(result){
```
`addInitActionResponse()` takes in more arguments than this.

kungoyal replied:
Fixed

kungoyal replied:
This has been implemented similar to addXReflectResponse functions with data being an object with questionID as key and choiceID as value

rasdy reviewed (changes requested):
Found a couple issues

rlee287 committed:
Merge branch 'goonuniverse' into differential_team

sspathak committed:
q&a.js: fixed syntax error

kungoyal committed:
Fixed addXActionResponse functions

rlee287 committed:
Rewrite the action page questions and choices function

rlee287 committed:
Update app.js action page endpoints to handle new JSON object format

rlee287 committed:
More checks in SQL schema

rasdy commented on goon-universe/app.js:
```diff
@@ -226,7 +226,7 @@ router.route('/scenarios/initialAction')
         }
         else{
         db.getInitActionPageQuestionsAndChoices(scenarioID, function(result){
```
The action GET endpoints should return the question and choice IDs that the action PUT endpoint expects.

rlee287 replied:
This should be fixed by the latest commit.

rlee287 committed:
Return question and mcq choice IDs from action page functions

rlee287 committed:
Add addInitActionPage to module.exports

rlee287 committed:
Add issue importance weight to SQL schema

rlee287 committed:
Check studentID and questionID correctness in addMCQResponse

rlee287 committed:
Return status boolean for addMCQResponse to change callback correctly

rlee287 committed:
Handle questionID/responseID nonexistence in addMCQResponse

rlee287 committed:
Also check validity of choiceID in addMCQResponse

rasdy reviewed (approved):
Looks good
